File: proxy_switch/features/git.py
"""Git proxy configuration via git config --global.

Manages http.proxy, https.proxy, and http.*.noProxy in ~/.gitconfig.
"""

from __future__ import annotations

from typing import Dict, Optional, List

from ..core.models import Result, StatusInfo

NAME = "git"
DESCRIPTION = "Git VCS proxy (git config --global)"
CONFIG_FILES = ["~/.gitconfig"]


def _run_git(args: str, executor=None) -> dict:
    """Run a git config command and return {'success': bool, 'output': str}."""
    cmd = f"git config --global {args}"
    if executor:
        result = executor.run(cmd)
        return {"success": result.returncode == 0, "output": result.stdout.strip()}
    else:
        import subprocess
        try:
            r = subprocess.run(cmd, shell=True, capture_output=True, text=True)
            return {"success": r.returncode == 0, "output": r.stdout.strip()}
        except Exception as e:
            return {"success": False, "output": str(e)}


def detect(executor=None) -> bool:
    """Check if git is installed."""
    if executor:
        return executor.run("command -v git").returncode == 0
    import shutil
    return shutil.which("git") is not None


def enable(proxy_config: Dict[str, str], executor=None) -> Result:
    """Set git http.proxy, https.proxy, and http.*.noProxy."""
    http = proxy_config.get("http_proxy", "")
    https = proxy_config.get("https_proxy", "")
    no_proxy = proxy_config.get("no_proxy", "")

    errors = []
    if http:
        r = _run_git(f'http.proxy "{http}"', executor)
        if not r["success"]:
            errors.append(f"http.proxy: {r['output']}")
    if https:
        r = _run_git(f'https.proxy "{https}"', executor)
        if not r["success"]:
            errors.append(f"https.proxy: {r['output']}")
    if no_proxy:
        r = _run_git(f'http."*".noProxy "{no_proxy}"', executor)
        if not r["success"]:
            errors.append(f"noProxy: {r['output']}")

    if errors:
        return Result(success=False, message="; ".join(errors))
    return Result(success=True, message="Git proxy configured")


def disable(executor=None) -> Result:
    """Unset git proxy settings."""
    _run_git("--unset http.proxy", executor)
    _run_git("--unset https.proxy", executor)
    _run_git("--unset http.*.noProxy", executor)
    return Result(success=True, message="Git proxy disabled")


def status(executor=None) -> StatusInfo:
    """Check git proxy settings."""
    r1 = _run_git("--get http.proxy", executor)
    r2 = _run_git("--get https.proxy", executor)
    proxy = r1["output"] or r2["output"] or None

    return StatusInfo(
        enabled=bool(r1["output"] or r2["output"]),
        proxy=proxy,
        config_file="~/.gitconfig",
    )


def validate(executor=None) -> List[str]:
    """Check Git configuration for issues."""
    issues = []
    if not detect(executor):
        issues.append("Git is not installed on this system")
    return issues

File: src-tauri/src/proxy/remote/git.rs
use crate::models::{ComponentId, ManualStep, OpResult, ProxyConfig, ProxyStatus};
use crate::proxy::ProxyModule;
use crate::ssh::connection::SshSession;

pub struct GitRemoteModule;

impl ProxyModule for GitRemoteModule {
    fn name(&self) -> &'static str {
        "git_remote"
    }
    fn description(&self) -> &'static str {
        "Git VCS proxy (git config --global)"
    }
    fn config_files(&self) -> Vec<String> {
        vec!["~/.gitconfig".into()]
    }
    fn manual_steps(&self) -> Vec<(String, Vec<String>)> {
        vec![(
            "Configure Git proxy".into(),
            vec![
                "git config --global http.proxy http://your-proxy:port".into(),
                "git config --global https.proxy http://your-proxy:port".into(),
                String::new(),
                "# To remove:".into(),
                "git config --global --unset http.proxy".into(),
                "git config --global --unset https.proxy".into(),
                String::new(),
                "# Config file: ~/.gitconfig".to_string(),
            ],
        )]
    }
    fn detect(&self, session: &SshSession) -> bool {
        session.tool_exists("git")
    }
    fn status(&self, session: &SshSession) -> ProxyStatus {
        let http = git_config_get(session, "http.proxy");
        let https = git_config_get(session, "https.proxy");
        let _proxy = if !https.is_empty() {
            Some(https.clone())
        } else if !http.is_empty() {
            Some(http.clone())
        } else {
            None
        };
        ProxyStatus {
            component: ComponentId::GitRemote,
            installed: self.detect(session),
            enabled: !http.is_empty() || !https.is_empty(),
            current_http_proxy: if http.is_empty() { None } else { Some(http) },
            current_https_proxy: if https.is_empty() { None } else { Some(https) },
            current_no_proxy: None,
            current_mirror: None,
            config_files: self.config_files(),
            manual_setup_steps: self
                .manual_steps()
                .into_iter()
                .map(|(t, c)| ManualStep {
                    title: t,
                    commands: c,
                })
                .collect(),
        }
    }
    fn enable(&self, session: &SshSession, config: &ProxyConfig) -> OpResult {
        let mut errors = Vec::new();
        if !config.http_proxy.is_empty() {
            let r = session.run(
                &format!("git config --global http.proxy \"{}\"", config.http_proxy),
                10,
            );
            if r.exit_code != 0 {
                errors.push(format!("http.proxy: {}", r.stderr));
            }
        }
        if !config.https_proxy.is_empty() {
            let r = session.run(
                &format!(
                    "git config --global https.proxy \"{}\"",
                    config.https_proxy
                ),
                10,
            );
            if r.exit_code != 0 {
                errors.push(format!("https.proxy: {}", r.stderr));
            }
        }
        if errors.is_empty() {
            OpResult::ok("Git proxy configured")
        } else {
            OpResult::err(errors.join("; "))
        }
    }
    fn disable(&self, session: &SshSession) -> OpResult {
        session.run("git config --global --unset http.proxy 2>/dev/null", 10);
        session.run("git config --global --unset https.proxy 2>/dev/null", 10);
        session.run("git config --global --unset http.*.noProxy 2>/dev/null", 10);
        OpResult::ok("Git proxy disabled")
    }
}

fn git_config_get(session: &SshSession, key: &str) -> String {
    session
        .run(
            &format!("git config --global --get {} 2>/dev/null || echo ''", key),
            5,
        )
        .stdout
}
